File: fraud_block_agent_demo/encode_slideshow.m
#import <AppKit/AppKit.h>
#import <AVFoundation/AVFoundation.h>
#import <CoreVideo/CoreVideo.h>
#import <Foundation/Foundation.h>

static CVPixelBufferRef CreatePixelBuffer(NSURL *url, int width, int height) {
    NSImage *image = [[NSImage alloc] initWithContentsOfURL:url];
    if (!image) return NULL;
    NSRect proposed = NSMakeRect(0, 0, width, height);
    CGImageRef cgImage = [image CGImageForProposedRect:&proposed context:nil hints:nil];
    if (!cgImage) return NULL;

    NSDictionary *attrs = @{
        (NSString *)kCVPixelBufferCGImageCompatibilityKey: @YES,
        (NSString *)kCVPixelBufferCGBitmapContextCompatibilityKey: @YES,
    };
    CVPixelBufferRef buffer = NULL;
    CVReturn result = CVPixelBufferCreate(kCFAllocatorDefault, width, height,
                                           kCVPixelFormatType_32ARGB,
                                           (__bridge CFDictionaryRef)attrs, &buffer);
    if (result != kCVReturnSuccess || !buffer) return NULL;

    CVPixelBufferLockBaseAddress(buffer, 0);
    CGContextRef context = CGBitmapContextCreate(
        CVPixelBufferGetBaseAddress(buffer), width, height, 8,
        CVPixelBufferGetBytesPerRow(buffer), CGColorSpaceCreateDeviceRGB(),
        kCGImageAlphaNoneSkipFirst);
    if (!context) {
        CVPixelBufferUnlockBaseAddress(buffer, 0);
        CVPixelBufferRelease(buffer);
        return NULL;
    }
    CGContextSetRGBFillColor(context, 1, 1, 1, 1);
    CGContextFillRect(context, CGRectMake(0, 0, width, height));
    CGContextDrawImage(context, CGRectMake(0, 0, width, height), cgImage);
    CGContextRelease(context);
    CVPixelBufferUnlockBaseAddress(buffer, 0);
    return buffer;
}

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        if (argc != 3) {
            fprintf(stderr, "Usage: encode_slideshow <frames_dir> <output.mp4>\n");
            return 2;
        }
        NSString *framesPath = [NSString stringWithUTF8String:argv[1]];
        NSString *outputPath = [NSString stringWithUTF8String:argv[2]];
        NSFileManager *fm = [NSFileManager defaultManager];
        NSError *error = nil;
        NSArray<NSString *> *names = [fm contentsOfDirectoryAtPath:framesPath error:&error];
        if (!names) {
            fprintf(stderr, "Cannot list frames: %s\n", error.localizedDescription.UTF8String);
            return 3;
        }
        NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(NSString *name, NSDictionary *bindings) {
            return [name hasPrefix:@"scene_"] && [[name pathExtension].lowercaseString isEqualToString:@"png"];
        }];
        names = [[names filteredArrayUsingPredicate:predicate] sortedArrayUsingSelector:@selector(compare:)];

        NSString *durationsText = [NSString stringWithContentsOfFile:[framesPath stringByAppendingPathComponent:@"durations.txt"]
                                                            encoding:NSUTF8StringEncoding error:&error];
        if (!durationsText) {
            fprintf(stderr, "Cannot read durations: %s\n", error.localizedDescription.UTF8String);
            return 4;
        }
        NSMutableArray<NSNumber *> *durations = [NSMutableArray array];
        for (NSString *line in [durationsText componentsSeparatedByCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]]) {
            if (line.length) [durations addObject:@(line.integerValue)];
        }
        if (names.count == 0 || names.count != durations.count) {
            fprintf(stderr, "Frame and duration counts do not match.\n");
            return 5;
        }

        [fm removeItemAtPath:outputPath error:nil];
        NSURL *outputURL = [NSURL fileURLWithPath:outputPath];
        AVAssetWriter *writer = [[AVAssetWriter alloc] initWithURL:outputURL fileType:AVFileTypeMPEG4 error:&error];
        if (!writer) {
            fprintf(stderr, "Cannot create writer: %s\n", error.localizedDescription.UTF8String);
            return 6;
        }
        const int width = 1920, height = 1080;
        const int32_t fps = 30;
        NSDictionary *compression = @{
            AVVideoAverageBitRateKey: @5000000,
            AVVideoProfileLevelKey: AVVideoProfileLevelH264HighAutoLevel,
        };
        NSDictionary *settings = @{
            AVVideoCodecKey: AVVideoCodecTypeH264,
            AVVideoWidthKey: @(width),
            AVVideoHeightKey: @(height),
            AVVideoCompressionPropertiesKey: compression,
        };
        AVAssetWriterInput *input = [[AVAssetWriterInput alloc] initWithMediaType:AVMediaTypeVideo outputSettings:settings];
        input.expectsMediaDataInRealTime = NO;
        NSDictionary *sourceAttrs = @{
            (NSString *)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_32ARGB),
            (NSString *)kCVPixelBufferWidthKey: @(width),
            (NSString *)kCVPixelBufferHeightKey: @(height),
        };
        AVAssetWriterInputPixelBufferAdaptor *adaptor =
            [[AVAssetWriterInputPixelBufferAdaptor alloc] initWithAssetWriterInput:input
                                                        sourcePixelBufferAttributes:sourceAttrs];
        if (![writer canAddInput:input]) {
            fprintf(stderr, "Cannot add video input.\n");
            return 7;
        }
        [writer addInput:input];
        if (![writer startWriting]) {
            fprintf(stderr, "Cannot start writer: %s\n", writer.error.localizedDescription.UTF8String);
            return 8;
        }
        [writer startSessionAtSourceTime:kCMTimeZero];

        int64_t frameIndex = 0;
        for (NSUInteger scene = 0; scene < names.count; scene++) {
            NSURL *imageURL = [NSURL fileURLWithPath:[framesPath stringByAppendingPathComponent:names[scene]]];
            CVPixelBufferRef buffer = CreatePixelBuffer(imageURL, width, height);
            if (!buffer) {
                fprintf(stderr, "Cannot create pixel buffer for %s\n", imageURL.path.UTF8String);
                return 9;
            }
            NSInteger repeat = durations[scene].integerValue * fps;
            for (NSInteger i = 0; i < repeat; i++) {
                while (!input.readyForMoreMediaData) [NSThread sleepForTimeInterval:0.005];
                CMTime time = CMTimeMake(frameIndex, fps);
                if (![adaptor appendPixelBuffer:buffer withPresentationTime:time]) {
                    fprintf(stderr, "Append failed: %s\n", writer.error.localizedDescription.UTF8String);
                    CVPixelBufferRelease(buffer);
                    return 10;
                }
                frameIndex++;
            }
            CVPixelBufferRelease(buffer);
            printf("Encoded scene %lu/%lu\n", (unsigned long)(scene + 1), (unsigned long)names.count);
        }

        [input markAsFinished];
        dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
        [writer finishWritingWithCompletionHandler:^{ dispatch_semaphore_signal(semaphore); }];
        dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
        if (writer.status != AVAssetWriterStatusCompleted) {
            fprintf(stderr, "Encoding failed: %s\n", writer.error.localizedDescription.UTF8String);
            return 11;
        }
        printf("Created %s\n", outputPath.UTF8String);
    }
    return 0;
}

File: fraud_block_agent_demo/encode_slideshow.swift
import AppKit
import AVFoundation
import CoreVideo
import Foundation

guard CommandLine.arguments.count == 3 else {
    fputs("Usage: swift encode_slideshow.swift <frames_dir> <output.mp4>\n", stderr)
    exit(2)
}

let framesDir = URL(fileURLWithPath: CommandLine.arguments[1], isDirectory: true)
let outputURL = URL(fileURLWithPath: CommandLine.arguments[2])
let fileManager = FileManager.default

let imageURLs = try fileManager.contentsOfDirectory(at: framesDir, includingPropertiesForKeys: nil)
    .filter { $0.lastPathComponent.hasPrefix("scene_") && $0.pathExtension.lowercased() == "png" }
    .sorted { $0.lastPathComponent < $1.lastPathComponent }

let durationsText = try String(contentsOf: framesDir.appendingPathComponent("durations.txt"))
let durations = durationsText.split(whereSeparator: \.isWhitespace).compactMap { Int($0) }

guard !imageURLs.isEmpty, imageURLs.count == durations.count else {
    fputs("Frame and duration counts do not match.\n", stderr)
    exit(3)
}

try? fileManager.removeItem(at: outputURL)

let width = 1920
let height = 1080
let fps: Int32 = 30
let writer = try AVAssetWriter(outputURL: outputURL, fileType: .mp4)
let settings: [String: Any] = [
    AVVideoCodecKey: AVVideoCodecType.h264,
    AVVideoWidthKey: width,
    AVVideoHeightKey: height,
    AVVideoCompressionPropertiesKey: [
        AVVideoAverageBitRateKey: 5_000_000,
        AVVideoProfileLevelKey: AVVideoProfileLevelH264HighAutoLevel
    ]
]
let input = AVAssetWriterInput(mediaType: .video, outputSettings: settings)
input.expectsMediaDataInRealTime = false
let attributes: [String: Any] = [
    kCVPixelBufferPixelFormatTypeKey as String: Int(kCVPixelFormatType_32ARGB),
    kCVPixelBufferWidthKey as String: width,
    kCVPixelBufferHeightKey as String: height
]
let adaptor = AVAssetWriterInputPixelBufferAdaptor(assetWriterInput: input, sourcePixelBufferAttributes: attributes)
guard writer.canAdd(input) else {
    fputs("Cannot add video input.\n", stderr)
    exit(4)
}
writer.add(input)

func makePixelBuffer(from url: URL) -> CVPixelBuffer? {
    guard let nsImage = NSImage(contentsOf: url) else { return nil }
    var proposed = CGRect(x: 0, y: 0, width: width, height: height)
    guard let cgImage = nsImage.cgImage(forProposedRect: &proposed, context: nil, hints: nil) else { return nil }

    var buffer: CVPixelBuffer?
    let status = CVPixelBufferCreate(
        kCFAllocatorDefault,
        width,
        height,
        kCVPixelFormatType_32ARGB,
        attributes as CFDictionary,
        &buffer
    )
    guard status == kCVReturnSuccess, let pixelBuffer = buffer else { return nil }
    CVPixelBufferLockBaseAddress(pixelBuffer, [])
    defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, []) }

    guard let context = CGContext(
        data: CVPixelBufferGetBaseAddress(pixelBuffer),
        width: width,
        height: height,
        bitsPerComponent: 8,
        bytesPerRow: CVPixelBufferGetBytesPerRow(pixelBuffer),
        space: CGColorSpaceCreateDeviceRGB(),
        bitmapInfo: CGImageAlphaInfo.noneSkipFirst.rawValue
    ) else { return nil }

    context.setFillColor(NSColor.white.cgColor)
    context.fill(CGRect(x: 0, y: 0, width: width, height: height))
    context.translateBy(x: 0, y: CGFloat(height))
    context.scaleBy(x: 1, y: -1)
    context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
    return pixelBuffer
}

guard writer.startWriting() else {
    fputs("Could not start writer: \(writer.error?.localizedDescription ?? "unknown error")\n", stderr)
    exit(5)
}
writer.startSession(atSourceTime: .zero)

var frameIndex: Int64 = 0
for (sceneIndex, imageURL) in imageURLs.enumerated() {
    guard let pixelBuffer = makePixelBuffer(from: imageURL) else {
        fputs("Could not load \(imageURL.path)\n", stderr)
        exit(6)
    }
    let repeatedFrames = durations[sceneIndex] * Int(fps)
    for _ in 0..<repeatedFrames {
        while !input.isReadyForMoreMediaData {
            Thread.sleep(forTimeInterval: 0.005)
        }
        let time = CMTime(value: frameIndex, timescale: fps)
        guard adaptor.append(pixelBuffer, withPresentationTime: time) else {
            fputs("Append failed: \(writer.error?.localizedDescription ?? "unknown error")\n", stderr)
            exit(7)
        }
        frameIndex += 1
    }
    print("Encoded scene \(sceneIndex + 1)/\(imageURLs.count)")
}

input.markAsFinished()
let semaphore = DispatchSemaphore(value: 0)
writer.finishWriting { semaphore.signal() }
semaphore.wait()

guard writer.status == .completed else {
    fputs("Encoding failed: \(writer.error?.localizedDescription ?? "unknown error")\n", stderr)
    exit(8)
}

print("Created \(outputURL.path)")
